### ISSUE TITLE
Updating

### ISSUE THREAD

**Issue by SquidDev** [Team-CC-Corp/Grin-Get#3]
I don't think it is implemented - updating would be really nice. Something like `grin-get update [package name...]`. Leaving it blank would update everything, specifying packages would attempt to update them.

I'm happy to put together a PR, just would ask you first.

**Comment by ElvishJerricco** [comment]
Yea this is rather important. Grin-Get has a whole bunch of things that ought to be implemented. If I weren't so busy as well as focused on other projects, there's a lot I would do with it.

**Comment by SquidDev** [comment]
I'm just seeing a bit more interest in ClamShell and think it would be nice if Grin-Get were a bit more fully-featured. I'll put something together in the next week or so.

**Comment by ElvishJerricco** [comment]
True. I need to start looking at expanding the Bish language for ClamShell. I read up on [monadic parsing](http://eprints.nottingham.ac.uk/223/1/pearl.pdf) in haskell (which is *fascinating*). I'd like to redesign the parser to be as monadic as Lua can concisely be, and I'd like to use the improved parser to implement some desirable language features.

**Comment by SquidDev** [comment]
Looks pretty cool! The composition element reminds me of [Sprache](https://github.com/sprache/Sprache) and [LPeg](http://www.inf.puc-rio.br/~roberto/lpeg/). I think Moonscript's parser is powered by LPeg, and I have considered using it for ClamShell, though monadic parsing would be more powerful.

**Comment by ElvishJerricco** [comment]
I actually went ahead and got started on a [library](https://github.com/Team-CC-Corp/Parsel) last night =P
```lua
parsel.symbol"local":bind(function(a) -- a is the string "local"
    -- Haven't made a simple "identifier" parser yet...
    return parsel.satisfy(function(c) return c:find("%a") ~= nil end)
    :many()
    :token() -- make it consume trailing whitespace
    :sepBy1(parsel.symbol",") -- parse multiple (at least one) instance, separated by comma
    :bind(function(names)
        -- names is the array of parsed names
        return parsel.symbol("="):bind(function(equals)
            return someExpressionParser()
        end)
    end)
end)
```
That will parse Lua's local declarations (excluding the expression parser). It's a bit cumbersome to use the bind operator without Haskell's do notation, but I'm looking into workarounds. For example, something like this might make it easier.
```lua
bindMany(startingMonad,
function(startsReturn) return getter end,
function(thatsReturn) return otherGetter end,
function(all, three, returns)
    -- perform computation on all three gotten values
    return computation
end)

-- Also the >> operator in haskell
someParser:sequence(otherParser)
-- equivalent to
someParser:bind(function(a) return otherParser end)
```

**Comment by SquidDev** [comment]
One thing you could do would be two wrap the functions with metatables to add custom operators. So for instance `a .. b` could be parse `a` then `b` (a sequence, from what I can tell the `>>=` operator) and then `a + b` would be the same but without preserving the value of `a`?

I'm probably just missing the point of moands here. My knowledge of functional programming starts and stops with lambdas.

**Comment by ElvishJerricco** [comment]
Haha yea you're missing the purpose of bind (>>=) but you've got the other bind operator (>>) correct.
```haskell
a >>= \x -> return (someOp x)
```

Note that `return` is a function, not a language keyword. It takes a value and wraps it in the contextual monad. The parsel equivalent is `from`, since I can't use the `return` keyword. Bind unwraps the value in a and calls the function with that as the parameter. Then bind returns something based on the return of that function (often just whatever that function returned exactly).

So it's not `a .. b`, it's `a .. function(a) return from(someOp(a)) end` The `>>` operator can be done that way though, since it basically means this

```haskell
a >> b = a >>= \x -> b
```

meaning that it's just returning `b` without any regard for `x`.

**Comment by SquidDev** [comment]
Having read some article I think I understand. In other news: I rewrote your test thing in a slightly different way:

```lua
local a = parsel.symbol "a"
local abtest = (a/"b"/"test"):many()
local braces = ("{" .. abtest .. "}") % function(obj) return obj[2][1] end
local p = abtest .. braces

local result = p(...)
```

The `/` symbol is aliased to `:otherwise`, `+` to `:bind`, `%` to `:fmap`.

`-` is also aliased to `:discardBind`, though that I'm less sure about. As you can see the `..` operator isn't great:

```lua
"{" .. func .. "}"
-- Results in
{ "{", { result, "}" } }
```
Each operator wraps the previous two, which isn't great. You can simplify with:
```lua
("{" - abtest .. "}") % function(obj) return obj[1] end
```

I'm not suggesting you use it - though I do like the `/` and `%` features.
